File: typescript/examples/playground-api/modules/docs.ts
import type { Express, Request, Response as ExpressResponse } from 'express'
import { readdir, readFile, stat } from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
// docs/api lives at the repo root, two levels above server/modules/.
const DOCS_ROOT = path.resolve(__dirname, '..', '..', '..', 'docs', 'api')

const LANGS = ['typescript', 'rust', 'go', 'python', 'ruby', 'php', 'lua', 'kotlin', 'swift'] as const
type Lang = (typeof LANGS)[number]

interface TreeNode {
  name: string
  path: string
  type: 'file' | 'dir'
  children?: TreeNode[]
}

async function buildTree(absDir: string, relDir = ''): Promise<TreeNode[]> {
  const entries = await readdir(absDir, { withFileTypes: true })
  const nodes: TreeNode[] = []
  for (const entry of entries) {
    if (entry.name.startsWith('.') || entry.name === 'node_modules') continue
    const relPath = relDir ? `${relDir}/${entry.name}` : entry.name
    const absPath = path.join(absDir, entry.name)
    if (entry.isDirectory()) {
      nodes.push({
        name: entry.name,
        path: relPath,
        type: 'dir',
        children: await buildTree(absPath, relPath),
      })
    } else if (entry.isFile() && entry.name.endsWith('.md')) {
      nodes.push({ name: entry.name, path: relPath, type: 'file' })
    }
  }
  // Folders first, then files; both alpha.
  nodes.sort((a, b) => {
    if (a.type !== b.type) return a.type === 'dir' ? -1 : 1
    return a.name.localeCompare(b.name)
  })
  return nodes
}

/** Reject any path that escapes the language root. */
function safeJoin(root: string, rel: string): string | null {
  const joined = path.resolve(root, rel)
  const rel2 = path.relative(root, joined)
  if (rel2.startsWith('..') || path.isAbsolute(rel2)) return null
  return joined
}

export function registerDocs(app: Express): void {
  // GET /api/v1/docs — which languages have generated docs?
  app.get('/api/v1/docs', async (_req: Request, res: ExpressResponse) => {
    const available: Record<Lang, boolean> = Object.fromEntries(LANGS.map((l) => [l, false])) as Record<Lang, boolean>
    for (const lang of LANGS) {
      const readme = path.join(DOCS_ROOT, lang, 'README.md')
      try {
        await stat(readme)
        available[lang] = true
      } catch {
        /* not generated yet */
      }
    }
    res.json({ root: DOCS_ROOT, available })
  })

  // GET /api/v1/docs/:lang/tree — file tree under docs/api/:lang/
  app.get('/api/v1/docs/:lang/tree', async (req: Request, res: ExpressResponse) => {
    const lang = String(req.params.lang)
    if (!(LANGS as readonly string[]).includes(lang)) {
      res.status(404).json({ error: 'unknown_lang' })
      return
    }
    const root = path.join(DOCS_ROOT, lang)
    try {
      await stat(root)
    } catch {
      res.status(404).json({ error: 'not_generated', hint: `Run: just docs-${recipeSlug(lang)}` })
      return
    }
    try {
      const tree = await buildTree(root)
      res.json({ lang, tree })
    } catch (err) {
      res.status(500).json({ error: 'tree_failed', detail: errMessage(err) })
    }
  })

  // GET /api/v1/docs/:lang/file?path=foo/bar.md — raw markdown content
  app.get('/api/v1/docs/:lang/file', async (req: Request, res: ExpressResponse) => {
    const lang = String(req.params.lang)
    if (!(LANGS as readonly string[]).includes(lang)) {
      res.status(404).json({ error: 'unknown_lang' })
      return
    }
    const rel = String(req.query.path ?? 'README.md')
    const root = path.join(DOCS_ROOT, lang)
    const abs = safeJoin(root, rel)
    if (!abs) {
      res.status(400).json({ error: 'unsafe_path' })
      return
    }
    if (!abs.endsWith('.md')) {
      res.status(400).json({ error: 'not_markdown' })
      return
    }
    try {
      const content = await readFile(abs, 'utf-8')
      res.type('text/markdown').send(content)
    } catch (err) {
      res.status(404).json({ error: 'not_found', detail: errMessage(err) })
    }
  })
}

function recipeSlug(lang: string): string {
  switch (lang) {
    case 'typescript':
      return 'ts'
    case 'rust':
      return 'rs'
    case 'python':
      return 'py'
    case 'ruby':
      return 'rb'
    case 'kotlin':
      return 'kt'
    default:
      return lang
  }
}

function errMessage(err: unknown): string {
  return err instanceof Error ? err.message : String(err)
}

File: go/examples/playground-api/docs.go
package main

// Serves the generated API reference markdown from <repo-root>/docs/api,
// with a path-escape guard. Override the root with the DOCS_ROOT env var
// when running the binary outside the repository checkout.

import (
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// docLangs are the languages the playground docs browser knows about.
var docLangs = []string{"typescript", "rust", "go", "python", "ruby", "php", "lua", "kotlin", "swift"}

// docsTreeNode is one entry of the docs file tree.
type docsTreeNode struct {
	// Name is the base name of the file or directory (e.g. "README.md").
	Name string `json:"name"`
	// Path is the slash-separated path relative to the language docs root;
	// the web app passes it back as the ?path= query of the file endpoint.
	Path string `json:"path"`
	// Type is "dir" for directories and "file" for markdown files; dirs
	// sort before files within a level.
	Type string `json:"type"`
	// Children holds the directory's child nodes; omitted for files.
	Children []docsTreeNode `json:"children,omitempty"`
}

// docsRoot resolves the generated-docs directory.
func docsRoot(repoRoot string) string {
	if override := os.Getenv("DOCS_ROOT"); override != "" {
		return override
	}
	if repoRoot == "" {
		return ""
	}
	return filepath.Join(repoRoot, "docs", "api")
}

// registerDocs mounts the generated-docs browsing endpoints.
func registerDocs(mux *http.ServeMux, a *app) {
	root := docsRoot(a.repoRoot)

	mux.HandleFunc("GET /api/v1/docs", func(w http.ResponseWriter, _ *http.Request) {
		available := map[string]bool{}
		for _, lang := range docLangs {
			_, err := os.Stat(filepath.Join(root, lang, "README.md"))
			available[lang] = err == nil
		}
		writeJSON(w, http.StatusOK, map[string]any{"root": root, "available": available})
	})

	mux.HandleFunc("GET /api/v1/docs/{lang}/tree", func(w http.ResponseWriter, r *http.Request) {
		lang := r.PathValue("lang")
		if !isDocLang(lang) {
			writeJSONError(w, http.StatusNotFound, "unknown_lang")
			return
		}
		langRoot := filepath.Join(root, lang)
		if _, err := os.Stat(langRoot); err != nil {
			writeJSON(w, http.StatusNotFound, map[string]string{
				"error": "not_generated",
				"hint":  "Run: just docs-" + docsRecipeSlug(lang),
			})
			return
		}
		tree, err := buildDocsTree(langRoot, "")
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{
				"error":  "tree_failed",
				"detail": err.Error(),
			})
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"lang": lang, "tree": tree})
	})

	mux.HandleFunc("GET /api/v1/docs/{lang}/file", func(w http.ResponseWriter, r *http.Request) {
		lang := r.PathValue("lang")
		if !isDocLang(lang) {
			writeJSONError(w, http.StatusNotFound, "unknown_lang")
			return
		}
		rel := r.URL.Query().Get("path")
		if rel == "" {
			rel = "README.md"
		}
		langRoot := filepath.Join(root, lang)
		abs, ok := safeJoin(langRoot, rel)
		if !ok {
			writeJSONError(w, http.StatusBadRequest, "unsafe_path")
			return
		}
		if !strings.HasSuffix(abs, ".md") {
			writeJSONError(w, http.StatusBadRequest, "not_markdown")
			return
		}
		content, err := os.ReadFile(abs)
		if err != nil {
			writeJSON(w, http.StatusNotFound, map[string]string{
				"error":  "not_found",
				"detail": err.Error(),
			})
			return
		}
		w.Header().Set("Content-Type", "text/markdown; charset=utf-8")
		_, _ = w.Write(content)
	})
}

// isDocLang reports whether lang is a known docs language.
func isDocLang(lang string) bool {
	for _, known := range docLangs {
		if lang == known {
			return true
		}
	}
	return false
}

// buildDocsTree walks the language docs directory: folders first, then
// markdown files, both alphabetical, skipping dotfiles and node_modules.
func buildDocsTree(absDir, relDir string) ([]docsTreeNode, error) {
	entries, err := os.ReadDir(absDir)
	if err != nil {
		return nil, err
	}
	nodes := []docsTreeNode{}
	for _, entry := range entries {
		name := entry.Name()
		if strings.HasPrefix(name, ".") || name == "node_modules" {
			continue
		}
		relPath := name
		if relDir != "" {
			relPath = relDir + "/" + name
		}
		if entry.IsDir() {
			children, err := buildDocsTree(filepath.Join(absDir, name), relPath)
			if err != nil {
				return nil, err
			}
			nodes = append(nodes, docsTreeNode{Name: name, Path: relPath, Type: "dir", Children: children})
		} else if strings.HasSuffix(name, ".md") {
			nodes = append(nodes, docsTreeNode{Name: name, Path: relPath, Type: "file"})
		}
	}
	sort.SliceStable(nodes, func(i, j int) bool {
		if nodes[i].Type != nodes[j].Type {
			return nodes[i].Type == "dir"
		}
		return nodes[i].Name < nodes[j].Name
	})
	return nodes, nil
}

// safeJoin joins rel onto root and rejects any path escaping the root.
func safeJoin(root, rel string) (string, bool) {
	joined := filepath.Join(root, filepath.FromSlash(rel))
	relBack, err := filepath.Rel(root, joined)
	if err != nil || relBack == ".." || strings.HasPrefix(relBack, ".."+string(filepath.Separator)) {
		return "", false
	}
	return joined, true
}

// docsRecipeSlug maps a docs language to its justfile recipe suffix.
func docsRecipeSlug(lang string) string {
	switch lang {
	case "typescript":
		return "ts"
	case "rust":
		return "rs"
	case "python":
		return "py"
	case "ruby":
		return "rb"
	case "kotlin":
		return "kt"
	default:
		return lang
	}
}
